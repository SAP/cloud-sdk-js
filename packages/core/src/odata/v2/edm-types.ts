/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { EdmTypeCommon, ExclusiveEdmTypeV2 } from '../common';

/**
 * Allowed Edm types for OData v2.
 */
export type EdmTypeV2 = EdmTypeCommon | ExclusiveEdmTypeV2;

export { EdmTypeV2 as EdmType };
