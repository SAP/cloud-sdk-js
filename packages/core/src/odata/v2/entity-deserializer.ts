/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import {
  entityDeserializer,
  EntityDeserializer
} from '../common/entity-deserializer';
import { edmToTsV2 } from './payload-value-converter';
import { extractODataEtagV2 } from './extract-odata-etag';
import { getLinkedCollectionResult } from './request-builder/response-data-accessor';

/**
 * Entity deserializer instance for v2 entities.
 * See [[EntityDeserializerType]] for the provided methods.
 */
const deserializer: EntityDeserializer = entityDeserializer(
  edmToTsV2,
  extractODataEtagV2,
  getLinkedCollectionResult
);

export const deserializeEntityV2 = deserializer.deserializeEntity;
export const deserializeComplexTypeV2 = deserializer.deserializeComplexType;

export { deserializeEntityV2 as deserializeEntity };
export { deserializeComplexTypeV2 as deserializeComplexType };
