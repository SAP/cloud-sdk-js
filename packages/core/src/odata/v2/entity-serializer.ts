/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import {
  EntitySerializer,
  entitySerializer
} from '../common/entity-serializer';
import { tsToEdmV2 } from './payload-value-converter';

/**
 * Entity serializer instance for v2 entities.
 * See [[EntitySerializerType]] for the provided methods.
 */
const serializer: EntitySerializer = entitySerializer(tsToEdmV2);

export const serializeEntityV2 = serializer.serializeEntity;
export const serializeComplexTypeV2 = serializer.serializeComplexType;
export const serializeEntityNonCustomFieldsV2 =
  serializer.serializeEntityNonCustomFields;

export { serializeEntityV2 as serializeEntity };
export { serializeComplexTypeV2 as serializeComplexType };
export { serializeEntityNonCustomFieldsV2 as serializeEntityNonCustomFields };
