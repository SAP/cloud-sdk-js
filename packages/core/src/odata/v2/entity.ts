/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { EntityBase, Constructable } from '../common/entity';
import { CustomField } from './selectable/custom-field';

/**
 * Super class for all representations of OData entity types.
 */
export class Entity extends EntityBase {
  protected static customFieldSelector<EntityT extends EntityBase>(
    fieldName: string,
    entityConstructor: Constructable<EntityT>
  ): CustomField<EntityT> {
    return new CustomField(fieldName, entityConstructor);
  }

  readonly _oDataVersion: 'v2' = 'v2';
}
