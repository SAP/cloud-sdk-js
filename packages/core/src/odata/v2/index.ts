/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

export * from './edm-types';
export * from './entity';
export * from './entity-deserializer';
export * from './entity-serializer';
export * from './filter-functions';
export * from './odata-batch-change-set';
export * from './odata-batch-request-builder';
export * from './odata-batch-retrieve-request';
export * from './payload-value-converter';
export * from './request-builder';
export * from './selectable/custom-field';
export * from './uri-conversion';
export * from './legacy-request-configs';
export * from './extract-odata-etag';
export * from '../common';
