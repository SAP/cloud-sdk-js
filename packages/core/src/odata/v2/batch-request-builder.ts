/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { errorWithCause, MapType } from '@sap-cloud-sdk/util';
import { v4 as uuid } from 'uuid';
import {
  Destination,
  DestinationNameAndJwt,
  DestinationOptions,
  toDestinationNameUrl
} from '../../scp-cf';
import { Constructable, BatchResponse } from '../common';
import { MethodRequestBuilderBase } from '../common/request-builder/request-builder-base';
import { ODataBatchRequestConfig } from '../common/request/odata-batch-request-config';
import { EntityV2 } from './entity';
import {
  ODataBatchChangeSetV2,
  serializeBatchRequest
} from './batch-request-serializer';
import {
  CreateRequestBuilderV2,
  DeleteRequestBuilderV2,
  GetAllRequestBuilderV2,
  GetByKeyRequestBuilderV2,
  UpdateRequestBuilderV2
} from './request-builder';
import {
  detectNewLineSymbol,
  partitionBatchResponse,
  buildRetrieveOrErrorResponse,
  buildWriteResponses
} from './batch-response-parser';

/**
 * Create a batch request to invoke multiple requests as a batch. The batch request builder accepts retrieve requests, i. e. [[GetAllRequestBuilder | getAll]] and [[GetByKeyRequestBuilder | getByKey]] requests and change sets, which in turn can contain [[CreateRequestBuilder | create]], [[UpdateRequestBuilder | update]] or [[DeleteRequestBuilder | delete]] requests.
 * The retrieve and change sets will be excuted in order, while the order within a change set can vary.
 */
export class ODataBatchRequestBuilderV2 extends MethodRequestBuilderBase<
  ODataBatchRequestConfig
> {
  /**
   * Creates an instance of ODataBatchRequestBuilder.
   *
   * @param defaultServicePath - Service path
   * @param requests - An array of retrieve requests or change sets
   * @param entityToConstructorMap - A map that holds the entity type to constructor mapping
   */
  constructor(
    readonly defaultServicePath: string,
    readonly requests: (
      | ODataBatchChangeSetV2<
          | CreateRequestBuilderV2<EntityV2>
          | UpdateRequestBuilderV2<EntityV2>
          | DeleteRequestBuilderV2<EntityV2>
        >
      | GetAllRequestBuilderV2<EntityV2>
      | GetByKeyRequestBuilderV2<EntityV2>
    )[],
    readonly entityToConstructorMap: MapType<Constructable<EntityV2>>
  ) {
    super(new ODataBatchRequestConfig(defaultServicePath, uuid()));
    this.requestConfig.payload = serializeBatchRequest(this);
  }

  /**
   * Execute the given request and return the according promise. Please notice: The sub-requests may fail even the main request is successful.
   *
   * @param destination - Targeted destination on which the request is performed.
   * @param options - Options to employ when fetching destinations.
   * @returns Promise resolving to the requested data.
   */
  async execute(
    destination: Destination | DestinationNameAndJwt,
    options?: DestinationOptions
  ): Promise<BatchResponse[]> {
    return this.build(destination, options)
      .then(request => request.execute())
      .then(response =>
        buildResponses(
          partitionBatchResponse(response),
          this.entityToConstructorMap,
          detectNewLineSymbol(response.data)
        )
      )
      .catch(error =>
        Promise.reject(
          errorWithCause(
            `The batch request failed! The destination provided: ${toDestinationNameUrl(
              destination
            )} and the options: ${options}.`,
            error
          )
        )
      );
  }
}

function buildResponses(
  responses: string[],
  entityToConstructorMap: MapType<Constructable<EntityV2>>,
  lineBreak: string
): BatchResponse[] {
  return responses.map(r =>
    buildResponse(r, entityToConstructorMap, lineBreak)
  );
}

function buildResponse(
  response: string,
  entityToConstructorMap: MapType<Constructable<EntityV2>>,
  lineBreak: string
) {
  if (isChangeSet(response)) {
    return buildWriteResponses(response, entityToConstructorMap, lineBreak);
  }
  if (isRetrieveRequestOrError(response)) {
    return buildRetrieveOrErrorResponse(
      response,
      entityToConstructorMap,
      lineBreak
    );
  }
  throw Error(
    `The response: ${JSON.stringify(
      response
    )} is not a valid retrieve request or change set, because it does not contain the proper Content-Type.`
  );
}

function isChangeSet(response: string): boolean {
  return !!response.match(/Content-Type: multipart\/mixed;/);
}

function isRetrieveRequestOrError(response: string): boolean {
  return !!response.match(/Content-Type: application\/http/);
}

export { ODataBatchRequestBuilderV2 as ODataBatchRequestBuilder };
