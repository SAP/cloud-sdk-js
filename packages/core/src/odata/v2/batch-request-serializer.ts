/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import {
  serializeChangeSet,
  serializeRequest
} from '../common/request-builder/batch/batch-request-serializer';
import { BatchChangeSet } from '../common/request-builder/batch/batch-change-set';
import { EntityV2 } from './entity';
import {
  CreateRequestBuilderV2,
  DeleteRequestBuilderV2,
  GetAllRequestBuilderV2,
  GetByKeyRequestBuilderV2,
  UpdateRequestBuilderV2
} from './request-builder';

/**
 * @deprecated Since v1.30.0. Use [[serializeRequest]] instead.
 * Build a string as the request body of the retrieve request.
 * Below is an example of the generated body, where the two empty line are mandatory to make the request valid.
 * *** example starts ***
 * Content-Type: application/http
 * Content-Transfer-Encoding: binary
 *
 * GET /SomeUrl/API_BUSINESS_PARTNER/A_BusinessPartnerBank?$format=json&$top=1 HTTP/1.1
 *
 *
 * *** example ends ***
 * @param request - The request builder of the retrieve request.
 * @returns The request body.
 */
export function toBatchRetrieveBodyV2(
  request: GetAllRequestBuilderV2<EntityV2> | GetByKeyRequestBuilderV2<EntityV2>
): string {
  return serializeRequest(request);
}

/**
 * @deprecated Since v1.30.0. Use [[serializeChangeSet]] instead.
 * Serialize change set to string.
 * @param changeSet - Change set containing a collection of write operations.
 * @returns The serialized string representation of a change set.
 */
export function toBatchChangeSetV2<
  T extends
    | CreateRequestBuilderV2<EntityV2>
    | UpdateRequestBuilderV2<EntityV2>
    | DeleteRequestBuilderV2<EntityV2>
>(changeSet: BatchChangeSet<T>): string | undefined {
  return serializeChangeSet(changeSet);
}

export { toBatchChangeSetV2 as toBatchChangeSet };
export { toBatchRetrieveBodyV2 as toBatchRetrieveBody };
