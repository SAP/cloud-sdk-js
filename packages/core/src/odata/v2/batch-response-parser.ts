/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import { last, MapType, createLogger } from '@sap-cloud-sdk/util';
import { HttpResponse } from '../../http-client';
import { getHeaderValue } from '../../header-builder';
import {
  getSingleResult,
  isCollectionResult,
  getCollectionResult
} from '../v2/request-builder/response-data-accessor';
import {
  Constructable,
  ErrorResponse,
  ReadResponse,
  WriteResponses,
  WriteResponse
} from '../common';
import { EntityV2 } from './entity';
import { deserializeEntityV2 } from './entity-deserializer';

const logger = createLogger({
  package: 'core',
  messageContext: 'batch-response-parser'
});

/**
 * Detects the system dependent line break in a string.
 * @param str The string to check for line breaks. Should have at least two lines, otherwise an error will be thrown.
 * @returns The system dependent line break
 */
export function detectNewLineSymbol(str: string): string {
  if (str.includes('\r\n')) {
    return '\r\n';
  }
  if (str.includes('\n')) {
    return '\n';
  }
  throw new Error('Cannot detect line breaks in the batch response body.');
}

/**
 * Get the response body from the string representation of a response.
 * @param response String representation of a response.
 * @returns The response body as a one line string.
 */
export function getResponseBody(response: string): string {
  const newLineSymbol = detectNewLineSymbol(response);
  const lines = response.split(newLineSymbol);

  // A valid response should contain at least three lines, part id, empty line and response body.
  if (lines.length >= 3) {
    return lines[lines.length - 1];
  }

  throw Error(
    `Cannot parse batch subrequest response body. Expected at least three lines in the response, got ${lines.length}.`
  );
}

/**
 * Parse the string representation of response headers into an object
 * @param headersStr Header string representation
 * @returns The headers as an object.
 */
function parseHeaders(headersStr: string): Record<string, any> {
  const newLineSymbol = detectNewLineSymbol(headersStr);
  return headersStr.split(newLineSymbol).reduce((headers, line) => {
    const [key, value] = line.split(':');
    return { ...headers, [key]: value };
  }, {});
}

/**
 * Get the boundary from the content type header value.
 * @param contentType Value of the content type header
 * @returns The boundary or undefined if none is given.
 */
function getBoundary(contentType: string | undefined): string | undefined {
  if (contentType) {
    return last(contentType.split('boundary='));
  }
}

/**
 * Split a batch response into an array of sub responses for the retrieve requests and changesets.
 * @param response The raw HTTP response.
 * @returns A list of sub responses represented as strings.
 */
export function splitBatchResponse(response: HttpResponse): string[] {
  const body = response.data.trim();
  if (!body) {
    return [];
  }

  const boundary = getBoundary(
    getHeaderValue('content-type', response.headers)
  );

  if (!boundary) {
    throw new Error(
      'Could not parse batch response. Expected multipart response with boundary.'
    );
  }

  return splitResponse(body, boundary);
}

/**
 * Split a changeset (sub) response into an array of sub responses.
 * @param changeSetResponse The string representation of a change set response.
 * @returns A list of sub responses represented as strings.
 */
export function splitChangeSetResponse(changeSetResponse: string): string[] {
  const [changeSetResponseHeader] = changeSetResponse.split('--');

  const headers = parseHeaders(changeSetResponseHeader);
  const boundary = getBoundary(getHeaderValue('content-type', headers));

  if (!boundary) {
    throw Error('Cannot parse change set.');
  }

  return splitResponse(changeSetResponse, boundary);
}

/**
 * Split a string representation of a response into sub responses given its boundary.
 * @param response The string representation of the response to split.
 * @param boundary The boundary to split by.
 * @returns A list of sub responses represented as strings.
 */
export function splitResponse(response: string, boundary: string): string[] {
  const parts = response.split(`--${boundary}`).map(part => part.trim());

  if (parts.length >= 3) {
    return parts.slice(1, parts.length - 1);
  }

  throw new Error(
    'Could not parse batch response body. Expected at least two response boundaries.'
  );
}

/**
 * Parse the entity name from the metadata uri. This should be the `__metadata` property of a single entity in the response.
 * @param uri The URI to parse the entity name from
 * @returns The entity name.
 */
export function parseEntityNameFromMetadataUri(
  uri: string | undefined
): string {
  if (!uri) {
    throw new Error(
      `Could not retrieve entity name from metadata. URI was: '${uri}'.`
    );
  }
  const [pathBeforeQuery] = uri.split('?');
  const [pathBeforeKeys] = pathBeforeQuery.split('(');
  const uriParts = pathBeforeKeys.split('/');

  // Remove another part in case of a trailing slash
  const entityName = uriParts.pop() || uriParts.pop();
  if (!entityName) {
    throw Error(
      `Could not retrieve entity name from metadata. Unknown URI format. URI was: '${uri}'.`
    );
  }
  return entityName;
}

// version specific
/**
 * Retrieve the constructor for a specific single response body.
 * @param responseBody The body of a single response as an object.
 * @param entityToConstructorMap Mapping between entity names and their respective constructors.
 * @returns The constructor if found in the mapping, undefined otherwise.
 */
export function getConstructor(
  responseBody: Record<string, any>,
  entityToConstructorMap: MapType<Constructable<EntityV2>>
): Constructable<EntityV2> | undefined {
  const entityJson = isCollectionResult(responseBody)
    ? getCollectionResult(responseBody)[0]
    : getSingleResult(responseBody);

  const entityUri = entityJson?.__metadata?.uri;
  if (entityUri) {
    return entityToConstructorMap[parseEntityNameFromMetadataUri(entityUri)];
  }

  logger.warn('Could not parse constructor from response body.');
}

/**
 * Parse the HTTP code of response.
 * @param response String representation of the response.
 * @returns The HTTP code.
 */
export function parseHttpCode(response: string): number {
  const group = response.match(/HTTP\/\d\.\d (\d{3}).*?/);

  if (group) {
    return parseInt(group[1].toString());
  }

  throw new Error('Cannot parse http code of the response.');
}

/**
 * Create a function to transform the parsed response body to a list of entities of the given type or an error.
 * @param body The parsed JSON reponse body.
 * @returns A function to be used for transformation of the read response.
 */
function asReadResponse(body: any) {
  return <T extends EntityV2>(constructor: Constructable<T>): Error | T[] => {
    if (body.error) {
      return new Error(body.error);
    }
    if (isCollectionResult(body)) {
      return getCollectionResult(body).map(r =>
        deserializeEntityV2(r, constructor)
      );
    }
    return [deserializeEntityV2(getSingleResult(body), constructor)];
  };
}

/**
 * Create a function to transform the parsed response body to an entity of the given type.
 * @param body The parsed JSON reponse body.
 * @returns A function to be used for transformation of the write response.
 */
function asWriteResponse(body) {
  return <T extends EntityV2>(constructor: Constructable<T>) =>
    deserializeEntityV2(getSingleResult(body), constructor);
}

/**
 * Get the body from the given response and parse it to JSON.
 * @param response The string representation of a single response.
 * @returns The parsed JSON representation of the response body.
 */
function parseResponseBody(response: string): Record<string, any> {
  const responseBody = getResponseBody(response);
  if (responseBody) {
    try {
      return JSON.parse(responseBody);
    } catch (err) {
      logger.error(
        `Could not parse response body. Invalid JSON. Original Error: ${err}`
      );
    }
  }
  return {};
}

/**
 * Parse a single response to the according response object type.
 * @param response The string representation of a single response.
 * @param entityToConstructorMap A map that holds the entity type to constructor mapping
 * @returns The parsed response in the according response object representation.
 */
export function parseResponse(
  response: string,
  entityToConstructorMap: MapType<Constructable<EntityV2>>
): WriteResponse | ReadResponse | ErrorResponse {
  const responseData = {
    body: parseResponseBody(response),
    httpCode: parseHttpCode(response)
  };

  // retrieve response
  if (responseData.httpCode === 200) {
    return {
      ...responseData,
      type: getConstructor(responseData.body, entityToConstructorMap),
      as: asReadResponse(responseData.body),
      isSuccess: () => true
    };
  }

  // change set sub response
  if (responseData.httpCode === 201 || responseData.httpCode === 204) {
    return {
      ...responseData,
      type: getConstructor(responseData.body, entityToConstructorMap),
      as: asWriteResponse(responseData.body)
    };
  }

  // error response
  return { ...responseData, isSuccess: () => false };
}

/**
 * Parse the complete batch HTTP response.
 * @param batchResponse HTTP response.
 * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
 * @returns An array of parsed sub responses of the batch response.
 */
export function parseBatchResponse(
  batchResponse: HttpResponse,
  entityToConstructorMap: MapType<Constructable<EntityV2>>
): (ErrorResponse | ReadResponse | WriteResponses)[] {
  return splitBatchResponse(batchResponse).map(response => {
    const contentType = getHeaderValue('content-type', parseHeaders(response));

    // is change set response
    if (isMultipartContentType(contentType)) {
      return {
        responses: splitChangeSetResponse(response).map(
          subResponse =>
            parseResponse(subResponse, entityToConstructorMap) as WriteResponse
        ),
        isSuccess: () => true
      };
    }

    // is read or error response
    if (isHttpContentType(contentType)) {
      return parseResponse(response, entityToConstructorMap) as
        | ReadResponse
        | ErrorResponse;
    }

    throw Error(
      `Cannot parse batch response. Unknown subresponse 'Content-Type' header '${contentType}'.`
    );
  });
}

function isMultipartContentType(contentType: string): boolean {
  return contentType.trim().startsWith('multipart/mixed');
}

function isHttpContentType(contentType: string): boolean {
  return contentType.trim().startsWith('application/http');
}
