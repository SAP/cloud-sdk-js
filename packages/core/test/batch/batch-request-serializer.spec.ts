/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import {
  toBatchChangeSetV2,
  toBatchRetrieveBodyV2,
  toEtagHeaderValue,
  toRequestPayload
} from '../../src';
import { TestEntity } from '../test-util/test-services/v2/test-service';
import {
  buildTestEntity,
  createChangeSetWithFakeId
} from '../test-util/batch-test-util';

describe('batch request serializer', () => {
  let testEntity: TestEntity;

  beforeEach(() => {
    testEntity = buildTestEntity();
  });

  it('serializes getAll request', () => {
    expect(
      toBatchRetrieveBodyV2(TestEntity.requestBuilder().getAll())
    ).toMatchSnapshot();
  });

  it('serializes getAll request with custom headers', () => {
    expect(
      toBatchRetrieveBodyV2(
        TestEntity.requestBuilder()
          .getAll()
          .withCustomHeaders({ 'Custom-Header': 'custom' })
      )
    ).toMatchSnapshot();
  });

  it('serializes getByKey request', () => {
    const getByKeyRequest = TestEntity.requestBuilder().getByKey(
      'testId',
      'test'
    );
    expect(toBatchRetrieveBodyV2(getByKeyRequest)).toMatchSnapshot();
  });

  describe('toRequestPayload', () => {
    it('serializes create request', () => {
      const createRequest = TestEntity.requestBuilder().create(testEntity);
      expect(toRequestPayload(createRequest, 'testId')).toMatchSnapshot();
    });

    it('serializes update request', () => {
      const updateRequest = TestEntity.requestBuilder().update(testEntity);
      expect(toRequestPayload(updateRequest, 'testId')).toMatchSnapshot();
    });

    it('serializes update request using put', () => {
      const updateRequest = TestEntity.requestBuilder()
        .update(testEntity)
        .replaceWholeEntityWithPut();
      expect(toRequestPayload(updateRequest, 'testId')).toMatchSnapshot();
    });

    it('serializes delete request with entity', () => {
      const deleteRequest = TestEntity.requestBuilder().delete(testEntity);
      expect(toRequestPayload(deleteRequest, 'testId')).toMatchSnapshot();
    });

    it('serializes delete request with id', () => {
      const deleteRequest = TestEntity.requestBuilder().delete('test', 'test');
      expect(toRequestPayload(deleteRequest, 'testId')).toMatchSnapshot();
    });

    it('serializes delete request with eTag', () => {
      const deleteRequest = TestEntity.requestBuilder().delete(
        testEntity.setVersionIdentifier('eTag')
      );
      expect(toRequestPayload(deleteRequest, 'testId')).toMatchSnapshot();
    });
  });

  describe('toEtagHeaderValue', () => {
    it('does not retrieve eTag for update request without eTag', () => {
      const updateRequest = TestEntity.requestBuilder().update(testEntity);
      expect(toEtagHeaderValue(updateRequest)).toBeUndefined();
    });

    it('retrieves eTag for update request with eTag', () => {
      const updateRequest = TestEntity.requestBuilder().update(
        testEntity.setVersionIdentifier('eTag')
      );
      expect(toEtagHeaderValue(updateRequest)).toEqual('eTag');
    });

    it('retrieves wildcard eTag for update request when eTag is ignored', () => {
      const updateRequest = TestEntity.requestBuilder()
        .update(testEntity.setVersionIdentifier('eTag'))
        .ignoreVersionIdentifier();
      expect(toEtagHeaderValue(updateRequest)).toEqual('*');
    });

    it('does not retrieve eTag for delete request without eTag', () => {
      const deleteRequest = TestEntity.requestBuilder().delete(testEntity);
      expect(toEtagHeaderValue(deleteRequest)).toBeUndefined();
    });

    it('retrieves eTag for delete request with eTag', () => {
      const deleteRequest = TestEntity.requestBuilder().delete(
        testEntity.setVersionIdentifier('eTag')
      );
      expect(toEtagHeaderValue(deleteRequest)).toEqual('eTag');
    });

    it('retrieves wildcard eTag for delete request when eTag is ignored', () => {
      const deleteRequest = TestEntity.requestBuilder()
        .delete(testEntity)
        .ignoreVersionIdentifier();
      expect(toEtagHeaderValue(deleteRequest)).toEqual('*');
    });

    it('does not retrieve eTag for create request', () => {
      const createRequest = TestEntity.requestBuilder().create(testEntity);
      expect(toEtagHeaderValue(createRequest)).toBeUndefined();
    });
  });

  describe('toBatchChangeSetV2', () => {
    it('serializes change set with one operation', () => {
      const createRequest = TestEntity.requestBuilder().create(testEntity);
      expect(
        toBatchChangeSetV2(createChangeSetWithFakeId(createRequest))
      ).toMatchSnapshot();
    });

    it('serializes change set with multiple operations', () => {
      const updateRequest = TestEntity.requestBuilder().update(testEntity);
      const createRequest = TestEntity.requestBuilder().create(testEntity);
      expect(
        toBatchChangeSetV2(
          createChangeSetWithFakeId(updateRequest, createRequest)
        )
      ).toMatchSnapshot();
    });

    it('returns undefined for empty change set', () => {
      expect(toBatchChangeSetV2(createChangeSetWithFakeId())).toMatchSnapshot();
    });
  });
});
