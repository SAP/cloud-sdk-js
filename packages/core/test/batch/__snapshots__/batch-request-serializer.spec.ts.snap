// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batch request serializer serializeBatchRequest serializes payload for batch subrequests 1`] = `
"--batch_boundary
Content-Type: multipart/mixed; boundary=changeSet_boundary

--changeSet_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

POST /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"KeyPropertyGuid\\":\\"guidId\\",\\"KeyPropertyString\\":\\"strId\\",\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}

--changeSet_boundary--
--batch_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

GET /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$format=json HTTP/1.1
Content-Type: application/json
Accept: application/json


--batch_boundary
Content-Type: multipart/mixed; boundary=changeSet_boundary

--changeSet_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

PATCH /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId') HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}

--changeSet_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

DELETE /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId') HTTP/1.1
Content-Type: application/json
Accept: application/json



--changeSet_boundary--
--batch_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

GET /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId')?$format=json HTTP/1.1
Content-Type: application/json
Accept: application/json


--batch_boundary--
"
`;

exports[`batch request serializer serializeChangeSet serializes change set with multiple operations 1`] = `
"Content-Type: multipart/mixed; boundary=changeSet_boundary

--changeSet_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

PATCH /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId') HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}

--changeSet_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

POST /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"KeyPropertyGuid\\":\\"guidId\\",\\"KeyPropertyString\\":\\"strId\\",\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}

--changeSet_boundary--"
`;

exports[`batch request serializer serializeChangeSet serializes change set with one operation 1`] = `
"Content-Type: multipart/mixed; boundary=changeSet_boundary

--changeSet_boundary
Content-Type: application/http
Content-Transfer-Encoding: binary

POST /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"KeyPropertyGuid\\":\\"guidId\\",\\"KeyPropertyString\\":\\"strId\\",\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}

--changeSet_boundary--"
`;

exports[`batch request serializer serializeRequest serializes create request 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

POST /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"KeyPropertyGuid\\":\\"guidId\\",\\"KeyPropertyString\\":\\"strId\\",\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}
"
`;

exports[`batch request serializer serializeRequest serializes delete request with eTag 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

DELETE /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId') HTTP/1.1
Content-Type: application/json
Accept: application/json
If-Match: eTag


"
`;

exports[`batch request serializer serializeRequest serializes delete request with entity 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

DELETE /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId') HTTP/1.1
Content-Type: application/json
Accept: application/json


"
`;

exports[`batch request serializer serializeRequest serializes delete request with id 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

DELETE /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'test',KeyPropertyString='test') HTTP/1.1
Content-Type: application/json
Accept: application/json


"
`;

exports[`batch request serializer serializeRequest serializes getAll request 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

GET /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$format=json HTTP/1.1
Content-Type: application/json
Accept: application/json

"
`;

exports[`batch request serializer serializeRequest serializes getAll request with custom headers 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

GET /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$format=json HTTP/1.1
Content-Type: application/json
Accept: application/json
Custom-Header: custom

"
`;

exports[`batch request serializer serializeRequest serializes getByKey request 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

GET /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'testId',KeyPropertyString='test')?$format=json HTTP/1.1
Content-Type: application/json
Accept: application/json

"
`;

exports[`batch request serializer serializeRequest serializes update request 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

PATCH /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId') HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}
"
`;

exports[`batch request serializer serializeRequest serializes update request using put 1`] = `
"Content-Type: application/http
Content-Transfer-Encoding: binary

PUT /sap/opu/odata/sap/API_TEST_SRV/A_TestEntity(KeyPropertyGuid=guid'guidId',KeyPropertyString='strId') HTTP/1.1
Content-Type: application/json
Accept: application/json

{\\"KeyPropertyGuid\\":\\"guidId\\",\\"KeyPropertyString\\":\\"strId\\",\\"Int16Property\\":12,\\"DoubleProperty\\":14.5}
"
`;
